### DISCUSSION TITLE
How would I go about using additional states with AsyncNotifier?

### DISCUSSION THREAD

**Discussion by joakimunge** [rrousselGit/riverpod#2401] [Q&A]
Hello once again!

AsyncNotifier currently allows three states `loading, data, error`. In my current application we would like to have a fourth state, for example `loadingMore`, to retain the current data state and show a loading indicator which is not the same as the loading indicator showed during the initial `loading` state.

How would I go about doing this, or is this outside the scope of an `AsyncNotifier` and I should instead use a `Notifier` with a custom state?

Appreciate any input you might have

**Comment by rrousselGit** [comment]
Rather than a custom AsyncValue you could use composition.

Do something like AsyncValue<LoadingMore<Data>>

**Reply by joakimunge** [reply]
Not sure I understand. This is the code in question. The AsyncValue type is a `Things`, so setting state to `AsyncValue<LoadingMore>` is not allowed.

Any chance you could elaborate? @rrousselGit 

```dart
class ThingsController extends AutoDisposeAsyncNotifier<Things> {
  @override
  FutureOr<Things> build() async {
    { ... } // Initial fetch
  }

  Future<void> getInitialThings() async {
    final thingsRepo = ref.read(thingsRepoProvider);
    state = const AsyncLoading();
    state = await AsyncValue.guard(
      () async {
        final things = await thingsRepo.getThings();
        return things;
      },
    );
  }

  Future<void> getNextThings(String startingAfter) async {
    final thingsRepo = ref.read(thingsRepoProvider);
    final previousState = state.requireValue;
    state = AsyncValue.data(previousState);
    
    state = // AsyncValue<LoadingMore> ?
        
    state = await AsyncValue.guard(
      () async {
        final nextThings =
            await thingsRepo.getNextThings(startingAfter);
        final things = [
          ...previousState.data.things,
          ...nextThings.data.things
        ];
        final data = nextThings.data.copyWith(things: things);
        return nextThings.copyWith(data: data);
      },
    );
  }
}
```
